### PR TITLE
Improve support of mbedtls_psa_get_random in client-only builds

### LINKED ISSUE #8825: Improve support of `mbedtls_psa_get_random` in client-only builds
Context: PSA has as standard RNG API that is accessible from everywhere as soon as `psa_crypto_init()` has been called. Legacy crypto APIs tend to accept `f_rng, p_rng` arguments because the legacy crypto API doesn't inlucde a globally-accessible RNG. The function `mbedtls_psa_get_random()` and companion macro `MBEDTLS_PSA_RANDOM_STATE` exist as a bridge between the two worlds, in that they allow legacy crypto APIs to use the PSA RNG.

However, as they are currently implemented, unless `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` is defined, they assume the RNG is built-in. This does work nicely with client-only (`CRYPTO_CLIENT && !CRYPTO_C`) builds. (And possibly neither with driver-only builds if the DRBG were accelerated.)

I suggest we:
- Change the implementation to be always the one currently use with `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG` . That it, only use the standard PSA APIs and quite making assumption about how it's implemented internally.
- Move the implementation from `psa_crypto.c` where (1) it doesn't belong (it's not a PSA API) and (2) it's guarded by `PSA_CRYPTO_C` as the rest of the file is, to `psa_util.c` where (1) it belongs (it's a PSA<->legacy bridge, declared in `psa_util.h`) and (2) it would be guarded only by `PSA_CRYPTO_CLIENT`.

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

Resolves #8825.

It does not strictly depends on #8842, but I suspect there might a conflict between the two.

## PR checklist

- [x] **changelog** done
- [x] **backport** not required
- [x] **tests** not required: not adding new code, just changing the way it is enabled

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
> I'm not sure about the redundant declarations. It seems to me we should just #include "mbedtls/psa_util.h" and be done with it, is there a reason not to? (Perhaps there was one before but not any more?)

IIRC the reason for the redundant declarations was that both the public include file and the private one needed different cases for the three underlying RNG implementations (CTR_DRBG, HMAC_DRBG or PSA external). Since those cases are going away, I suspect there's nothing particularly fragile anymore.

> So, do we need this file at all or can we just fold it into psa_crypto.c

(the file being `psa_crypto_random_impl.h`)

That's from https://github.com/Mbed-TLS/mbedtls/pull/3933. At the time `mbedtls/psa_util.c` already exposed the PSA RNG abstraction so `psa_crypto_rsa.c` and friends could have used it, but they didn't. I guess the idea was that the PSA implementation would directly depend on a PSA interface rather than an mbedtls interface exposing that abstraction? Anyway I think that's irrelevant now and only `psa_crypto.c` needs to include `psa_crypto_random_impl.h`.

I think `psa_crypto_random_impl.h` does need to remain a header file, because the tests of that abstraction (in `test_suite_psa_crypto_init`) need the declarations.

**Comment by valeriosetti** [comment]
I just checked and the only thing from `psa_crypto_random_impl.h` that it's referenced outside of `psa_crypto.c` is `MBEDTLS_PSA_HMAC_DRBG_MD_TYPE`. This means that most of the code from that header file can be moved to `psa_crypto.c`. *However* `psa_crypto.c` is already extremely long (> 8k lines), so that would make the file even longer and personally I'm not a big fan of this. Perhaps keeping the implementation in a separate file is a good solution.

**Comment by mpg** [comment]
> I think `psa_crypto_random_impl.h` does need to remain a header file, because the tests of that abstraction (in `test_suite_psa_crypto_init`) need the declarations.

I don't think they do. As I wrote earlier, I tried commenting out that line and the test suite was still building and passing in both the default and the full config.

**Comment by valeriosetti** [comment]
While writing the changelog for this PR I realized that I forgot to change the guard for `mbedtls_psa_get_random()` from `CRYPTO_C` to `CRYPTO_CLIENT`. However I suspect that this change might require the same fix for tests that was done in #8842. That's why I updated the description of the PR with this dependency.

**Comment by valeriosetti** [comment]
[This](https://github.com/Mbed-TLS/mbedtls/pull/8862/commits/1a58e9a2326d34254125c30ffad9acd33a610905) commit updates the guard from `CRYPTO_C` to `CRYPTO_CLIENT`. In order to solve test issues in `component_test_psa_crypto_client()`, a stub function is also added for client only build. This mimics what Gilles suggested in #8842 and therefore this is also the reason for which I expect a conflict once #8842 will be merged. However it should be quite trivial to solve.

**Review by mpg — DISMISSED** [review]
LGTM.

I still think `psa_crypto_random_impl.h` should be stripped down further, but that can be done later.

**Comment by valeriosetti** [comment]
> I still think `psa_crypto_random_impl.h` should be stripped down further, but that can be done later.

One question on this so that I can file up a new issue. As I anticipated in a [previous comment](https://github.com/Mbed-TLS/mbedtls/pull/8862#issuecomment-1968262003) most of the `psa_crypto_random_impl.h` file is not used anywhere a part from `psa_crypto.c` so we can move most of former's content in the latter at the price of latter's size increase.
Can I create a new C file to hold this part instead of moving everything to `psa_crypto.c`? This would mean that `inline` functions will loose their inlining feature, but this should keep `psa_crypto.c` more readable :shrug: . Wdyt?

File: ChangeLog.d/8825.txt
Comment (tom-daubney-arm) [inline ChangeLog.d/8825.txt:5]:
```suggestion
     used as random number generator function (f_rng) and context (p_rng) in
```

File: tests/src/psa_crypto_stubs.c
Comment (tom-daubney-arm) [inline tests/src/psa_crypto_stubs.c:25]:
```suggestion
#endif /* MBEDTLS_PSA_CRYPTO_CLIENT && !MBEDTLS_PSA_CRYPTO_C */
```

**Review by tom-daubney-arm — CHANGES_REQUESTED** [review]
Almost looks completely fine but with two very small nits that I almost feel guilty pulling you up on.

**Comment by gilles-peskine-arm** [comment]
> Can I create a new C file to hold this part instead of moving everything to psa_crypto.c? This would mean that inline functions will loose their inlining feature, but this should keep psa_crypto.c more readable 🤷 . Wdyt?

That tends to be a compromise between readability and code size. For example we split off `psa_crypto_driver_wrappers.c` to a separate `.o` file, then we moved some of it back into a header to allow the functions to be inlined. Link-time optimization would make this not a concern, but many projects that consume Mbed TLS can't easily use LTO.

**Review by tom-daubney-arm — APPROVED** [review]
LGTM, thanks for the changes.

**Review by mpg — APPROVED** [review]
LGTM

**Pull request merged**
